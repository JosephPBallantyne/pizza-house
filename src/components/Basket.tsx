import { useContext, useEffect, useState } from 'react';
import Alert from '@mui/material/Alert';
import AlertTitle from '@mui/material/AlertTitle';
import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import Dialog from '@mui/material/Dialog';
import Divider from '@mui/material/Divider';
import Grid from '@mui/material/Grid';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import BasketContext from '../context/basketProvider';
import { Pizza } from '../types/pizza';

const Basket: React.FC = () => {
  const extraToppingCost = 10;
  const { basketItems, setBasketItems } = useContext(BasketContext);
  const [basketChanged, setBasketChanged] = useState(false);
  const [checkout, setCheckout] = useState(false);
  let totalPrice = 0;
  basketItems.forEach((pizza: Pizza) => {
    totalPrice += pizza.price + pizza.extraToppings.length * extraToppingCost;
  });

  // useEffect(() => {
  //   const existingBasket = localStorage.getItem('basketItems');
  //   console.log(existingBasket);
  // }, []);

  const handleCheckout = () => {
    setCheckout(!checkout);
  };

  const handleRemovePizza = (i: number) => {
    basketItems.splice(i, 1);
    setBasketItems(basketItems);
    setBasketChanged(!basketChanged);
    localStorage.setItem('basketItems', JSON.stringify(basketItems));
  };

  return (
    <Card
      sx={{ maxWidth: 345, padding: 1, ml: 2, bgcolor: '#F3F6F9' }}
      variant="outlined"
    >
      <CardContent>
        <Typography gutterBottom variant="h6" component="div" sx={{ pb: 3 }}>
          Your Basket
        </Typography>
        {basketItems.length < 1 && (
          <Typography variant="body1" color="text.secondary" sx={{ pb: 1 }}>
            Your basket is empty
          </Typography>
        )}
        {basketItems.map((pizza, i) => {
          return (
            <Grid key={'basket' + i} container spacing={0}>
              <Grid key={i} xs={2} item>
                <IconButton
                  size="small"
                  edge="end"
                  color="inherit"
                  aria-label="delete"
                  sx={{ p: 0, m: 0 }}
                  onClick={() => handleRemovePizza(i)}
                >
                  <DeleteForeverIcon />
                </IconButton>
              </Grid>
              <Grid key={i} xs={8} item>
                <Typography variant="body1" color="text.primary">
                  {pizza.name} {pizza.size}&quot;
                </Typography>
              </Grid>
              <Grid
                key={'price' + i}
                xs={2}
                item
                sx={{ display: 'flex', justifyContent: 'flex-end' }}
              >
                ${pizza.price}
              </Grid>

              <Grid key={'extra' + i} xs={2} item></Grid>
              <Grid key={'extra' + i} xs={8} item>
                {pizza.extraToppings.length > 0 && (
                  <Typography variant="body2" color="text.secondary">
                    {pizza.extraToppings.length}x extra topping(s)
                  </Typography>
                )}
              </Grid>
              <Grid key={'priceExtra' + i} xs={2} item sx={{ pb: 3 }}>
                {pizza.extraToppings.length > 0 && (
                  <Typography
                    variant="body2"
                    color="text.secondary"
                    sx={{ display: 'flex', justifyContent: 'flex-end' }}
                  >
                    + ${pizza.extraToppings.length * extraToppingCost}
                  </Typography>
                )}
              </Grid>
            </Grid>
          );
        })}
        <Divider sx={{ py: 1 }} />
        <Typography variant="subtitle1" color="text.primary">
          <Grid
            container
            spacing={2}
            sx={{ display: 'flex', justifyContent: 'space-between', pt: 1 }}
          >
            <Grid item>Total</Grid>
            <Grid item>${totalPrice}</Grid>
          </Grid>
        </Typography>
      </CardContent>
      <CardActions>
        <Button
          size="small"
          variant="contained"
          disabled={basketItems.length < 1}
          onClick={handleCheckout}
        >
          Checkout
        </Button>
        <Dialog open={checkout} onBackdropClick={handleCheckout}>
          <Alert severity="success">
            <AlertTitle>Order Success</AlertTitle>
            Your pizza will be with you soon â€” <strong>sit tight!</strong>
          </Alert>
        </Dialog>
      </CardActions>
    </Card>
  );
};

export default Basket;
