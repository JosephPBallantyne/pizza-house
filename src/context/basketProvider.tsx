import React from 'react';
import { Pizza } from '../types/pizza';

const BasketContext = React.createContext<any>(null);

export const BasketContextProvider: React.FC<any> = ({
  children,
  value,
}: React.PropsWithChildren<{
  value: {
    basketItems: Pizza[];
    setBasketItems: React.Dispatch<React.SetStateAction<Pizza[]>>;
  };
}>) => {
  return (
    <BasketContext.Provider value={value}>{children}</BasketContext.Provider>
  );
};

export default BasketContext;
