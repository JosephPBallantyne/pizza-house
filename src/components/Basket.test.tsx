import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Basket from './Basket';
import '@testing-library/jest-dom/extend-expect';
import { BasketContextProvider } from '../context/basketProvider';

const BASKET =
  '[{"name":"Margherita","toppings":["Mozzarella Cheese","Tomato Sauce","Basil"],"price":78,"size":12,"extraToppings":[]}]';

describe('Basket component', () => {
  it('should render an empty basket', () => {
    expect.assertions(3);
    const { container } = render(
      <BasketContextProvider value={{ basketItems: [] }}>
        <Basket />
      </BasketContextProvider>,
      {}
    );
    expect(container).toHaveTextContent('Your basket is empty');
    expect(screen.getByRole('button', { name: /Checkout/i })).toBeDisabled();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a pizza in basket', () => {
    expect.assertions(4);
    const { container } = render(
      <BasketContextProvider value={{ basketItems: JSON.parse(BASKET) }}>
        <Basket />
      </BasketContextProvider>,
      {}
    );
    expect(container).not.toHaveTextContent('Your basket is empty');
    expect(container).toHaveTextContent('Margherita');
    expect(screen.getByRole('button', { name: /Checkout/i })).toBeEnabled();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should remove a pizza from the basket', async () => {
    expect.assertions(2);
    const { container } = render(
      <BasketContextProvider
        value={{ basketItems: JSON.parse(BASKET), setBasketItems: jest.fn() }}
      >
        <Basket />
      </BasketContextProvider>,
      {}
    );
    userEvent.click(screen.getByTestId('DeleteForeverIcon'));
    expect(screen.getByRole('button', { name: /Checkout/i })).toBeDisabled();
    expect(container).toHaveTextContent('Your basket is empty');
  });
});
