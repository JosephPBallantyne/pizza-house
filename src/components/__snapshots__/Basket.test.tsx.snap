// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basket component should render a pizza in basket 1`] = `
<div
  class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-ekvxvt-MuiPaper-root-MuiCard-root"
>
  <div
    class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <div
      class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-hxo9is-MuiTypography-root"
    >
      Your Basket
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-a0vo9r-MuiGrid-root"
      >
        <button
          aria-label="delete"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd MuiIconButton-sizeSmall css-b35x4v-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteForeverIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-12dpnmv-MuiGrid-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1gqyz35-MuiTypography-root"
        >
          Margherita
           
          12
          "
        </p>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-1ytmhr0-MuiGrid-root"
      >
        $
        78
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-a0vo9r-MuiGrid-root"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-12dpnmv-MuiGrid-root"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-1mhdmtd-MuiGrid-root"
      />
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-16x6gxp-MuiDivider-root"
    />
    <h6
      class="MuiTypography-root MuiTypography-subtitle1 css-65a2cr-MuiTypography-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1gisd3f-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          Total
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          $
          78
        </div>
      </div>
    </h6>
  </div>
  <div
    class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
  >
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root css-11qr2p8-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Checkout
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`Basket component should render an empty basket 1`] = `
<div
  class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-ekvxvt-MuiPaper-root-MuiCard-root"
>
  <div
    class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <div
      class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-hxo9is-MuiTypography-root"
    >
      Your Basket
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-12sdzeh-MuiTypography-root"
    >
      Your basket is empty
    </p>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-16x6gxp-MuiDivider-root"
    />
    <h6
      class="MuiTypography-root MuiTypography-subtitle1 css-65a2cr-MuiTypography-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1gisd3f-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          Total
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          $
          0
        </div>
      </div>
    </h6>
  </div>
  <div
    class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
  >
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root Mui-disabled css-11qr2p8-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Checkout
    </button>
  </div>
</div>
`;
